ballerina/variable_validator.bal

```diff
@@ -149,9 +168,65 @@ class VariableValidator {
             string message = string`Variable "$${<string> argument.getVariableName()}" got invalid value ` +
-            string`${value.toString()};Expected type ${getTypeName(argument)}. ${getTypeName(argument)}` +
+            string`${value.toString()};Expected type ${varDef.kind}. ${getTypeName(argument)}` +
             string` cannot represent value: ${value.toString()}`;
-            self.errors.push(getErrorDetailRecord(message, location)); 
+            self.errors.push(getErrorDetailRecord(message, argument.getLocation()));
+        }
+    }
+
+    isolated function checkVariableUsageCompatibility(__InputValue[] input, parser:VariableDefinition varDef,
+                                                      parser:ArgumentNode argNode) {
+        __InputValue? inputValue = getInputValueFromArray(input, argNode.getName());
+        __Type? actualType = self.getTypeRecordFromTypeName(varDef.kind);
+        if inputValue is __InputValue {
+            if actualType is __Type {
+                if inputValue.'type != actualType {
+                    if varDef?.defaultValue is () || getOfType(inputValue.'type)?.name != getOfType(actualType)?.name {
+                        string message = string`Variable "${<string>argNode.getVariableName()}" of type "${varDef.kind}" `+
+                        string`used in position expecting type "${getTypeNameFromType(inputValue.'type)}".`;
+                        self.errors.push(getErrorDetailRecord(message, argNode.getLocation()));
+                    }
+                }
+            } else {
+                string message = string`Variable "${<string>argNode.getVariableName()}" of type "${varDef.kind}" `+
+                string`used in position expecting type "${getTypeNameFromType(inputValue.'type)}".`;
+                self.errors.push(getErrorDetailRecord(message, argNode.getLocation()));
+            }
         }
     }
+
+    isolated function getTypeRecordFromTypeName(string typeName) returns __Type? {
+        string ofTypeName;
+        __Type wrapperType;
+        __Type? ofType;
+        if typeName.endsWith("!") {
+            wrapperType = {
+                kind: NON_NULL,
+                name: ()
+            };
+            ofTypeName = typeName.substring(0, typeName.length()-1);
+            ofType = self.getTypeRecordFromTypeName(ofTypeName);
+            if ofType is __Type {
+                wrapperType.ofType = ofType;
+            }
+            return wrapperType;
+        } else if typeName.startsWith("[") {
+            wrapperType = {
+                kind: LIST,
+                name: ()
+            };
+            ofTypeName = typeName.substring(1, typeName.length()-1);
+            ofType = self.getTypeRecordFromTypeName(ofTypeName);
+            if ofType is __Type {
+                wrapperType.ofType = ofType;
+            }
+            return wrapperType;
+        } else {
+            ofType = getTypeFromTypeArray(self.schema.types, typeName);
```

Comment:
What happens if the type is not found here?

Let's say we have `Person!` as the type and the `Person` object type is not found. This will return the following type:

```
{
    "name": (),
    "kind": NON_NULL,
    "ofType": () 
}
```

Which is wrong. We should return an error in this case.

Reply:
Since this is also checked at the `checkVariableUsageCompatibility( )`, error is not added to this function. Will change it.

Reply:
So are we traversing the same variable twice? Can't we handle all this within the same iteration?

As I mentioned in a previous comment, we are duplicating the condition as well right? Better if we can handle it here. (Or handle this there).